### PR TITLE
Updated to CheckStyle 8.22

### PR DESCRIPTION
**Pull request by lolgab**
- Bumped all dependencies to lastest version (except sbt-native-package)
- Added new rules using automatic importing from CheckStyle
- Fixed checkstyle documentation parser ( they added a link inside description for the first time, which broke our parser )

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -66,13 +68,15 @@ dockerBaseImage := "openjdk:8-jre-alpine"
...
-  case cmd@(Cmd("ADD", _)) => List(
+  case cmd @ Cmd("ADD", _) => Seq(
     Cmd("RUN", s"adduser -u 2004 -D $dockerUser"),
     cmd,
     Cmd("RUN", "mv /opt/docker/docs /docs")
   )
   case other => List(other)
 }
+
+s3region := Region.EU_Ireland
```
Comment (gjsduarte) [inline build.sbt]: Maybe a little comment explaining why is this necessary?

File: src/main/resources/docs/patterns.json [hunk 2]
```diff
@@ -368,7 +379,7 @@
       "default" : ".*"
     }, {
       "name" : "exceptionVariableName",
-      "default" : "^$"
+      "default" : "\"^$\" (empty)"
```
Comment (gjsduarte) [inline src/main/resources/docs/patterns.json]: This seems wrong. The `(empty)` part of the string should probably be removed

File: build.sbt [hunk 3]
```diff
@@ -19,10 +17,14 @@ lazy val toolVersionKey = settingKey[String]("The version of the underlying tool
...
 toolVersionKey := {
   val jsonFile = (resourceDirectory in Compile).value / "docs" / "patterns.json"
-  val toolMap = JSON.parseFull(Source.fromFile(jsonFile).getLines().mkString)
-    .getOrElse(throw new Exception("patterns.json is not a valid json"))
-    .asInstanceOf[Map[String, String]]
-  toolMap.getOrElse[String]("version", throw new Exception("Failed to retrieve 'version' from patterns.json"))
+  val source = Source.fromFile(jsonFile)
+  try {
+    val jsonString = source.getLines().mkString
+    val json = ujson.read(jsonString)
```
Comment (andreaTP) [inline build.sbt]: [sjson](https://github.com/eed3si9n/sjson-new) is already a transitive dependency of sbt itself, can we use it instead of another library for such a small use-case?

File: project/plugins.sbt [hunk 4]
```diff
@@ -1 +1,6 @@
-addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.3.6")
\ No newline at end of file
+resolvers += Resolver.sonatypeRepo("releases")
+addSbtPlugin("com.codacy" % "codacy-sbt-plugin" % "17.0.2")
+
+addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.3.15")
+
+libraryDependencies += "com.lihaoyi" %% "ujson" % "0.7.5" // to parse Json in the build.sbt file
```
Comment (andreaTP) [inline project/plugins.sbt]: as said, can we try to use `sjson`?

File: src/main/scala/codacy/checkstyle/CodacyListener.scala [hunk 5]
```diff
@@ -26,7 +26,7 @@ class CodacyListener extends AuditListener {
           Source.Line(event.getLine)
         )
       }
-  }
+  } catch { case NonFatal(e) => e.printStackTrace() }
```
Comment (andreaTP) [inline src/main/scala/codacy/checkstyle/CodacyListener.scala]: not sure is a good idea to print to std-out, maybe we can generate a tool failure with a proper error message

**Comment by lolgab** [comment]
Things to do: 
- [ ] Update `sbt-codacy` plugin and get rid of Region workaround.
- [ ] Make sjsonnew code for task `toolVersionKey` more idiomatic.
- [ ] Fix "\"^$\" (empty)" patterns.json (in the DocGenerator)

File: project/plugins.sbt (hunk 4 above)
Comment (rtfpessoa) [reply project/plugins.sbt]: Please no more JSON libraries in the code base

**Comment by rtfpessoa** [comment]
> a few minor comments, I'm wondering if it's possible to avoid to commit to Git the generated files and keep them in `resource_managed`.
> @rtfpessoa WDYT?

We use this files to check the contents of the docs. And specially since they are external and we cannot version them not having them in the repo seems like a bad idea.
We would never check their quality and we could have wierd outcomes during build

File: src/main/scala/codacy/checkstyle/CodacyListener.scala (hunk 5 above)
Comment (lolgab) [reply src/main/scala/codacy/checkstyle/CodacyListener.scala]: Refactored non handling the exception at all. So if there are problems we are aware of it.

File: project/plugins.sbt (hunk 4 above)
Comment (lolgab) [reply project/plugins.sbt]: Refactored using the already provided sjsonnew library.

File: build.sbt (hunk 3 above)
Comment (lolgab) [reply build.sbt]: Now I'm using the already provided sjsonnew library.

File: src/main/resources/docs/patterns.json (hunk 2 above)
Comment (lolgab) [reply src/main/resources/docs/patterns.json]: Now the DocGenerator is fixed and parses only the real regular expression and not the comments (like the `(empty)` you see here.

File: build.sbt (hunk 1 above)
Comment (lolgab) [reply build.sbt]: It's gone with new `codacy-sbt` update.

File: src/main/scala/codacy/checkstyle/DocGenerator.scala [hunk 6]
```diff
@@ -59,13 +59,14 @@ object DocGenerator {
...
                   val defaultValue = Option({
                     // Remove spaces and breaklines in default values
                     val defVal = default.text.replaceAll("""\n\s+""", "").trim.stripSuffix(".")
-                    // Remove quotes around default values
-                    if (defVal.startsWith("\"") && defVal.endsWith("\"")) {
-                      defVal.stripPrefix("\"").stripSuffix("\"")
+                    // Remove quotes around regular expressions
+                    if (tpe.text.trim == "Regular Expression" && defVal != "null") {
+                      // Leaves only what's inside outer quotes.
+                      defVal.dropWhileLeftRight(_ != '"').drop(1).dropRight(1)
```
Comment (andreaTP) [inline src/main/scala/codacy/checkstyle/DocGenerator.scala]: this implementation can easily fail, `drop` and `dropRight` are going to throw exceptions instead of avoiding to strip ( https://www.scala-lang.org/api/2.12.4/scala/collection/immutable/StringOps.html#stripSuffix(suffix:String):String )

File: src/main/scala/codacy/checkstyle/DocGenerator.scala [hunk 7]
```diff
@@ -150,4 +150,24 @@ object DocGenerator {
     }
   }
 
+
+  implicit class StringOps(val s: String) extends AnyVal {
```
Comment (andreaTP) [inline src/main/scala/codacy/checkstyle/DocGenerator.scala]: `StringOps` in Scala std lib is already pretty comprehensive, are we sure we can't avoid running our own implementation of StringOps? e.g. `takeWhile` have the same signature and in case a `reverse` will make the trick of removing from the other end ...

File: src/main/scala/codacy/checkstyle/DocGenerator.scala [hunk 8]
```diff
@@ -150,4 +150,24 @@ object DocGenerator {
...
+    /**
+      * Like Scala's dropWhile but it works on right side too.
+      *
+      * @param f a function to match chars to remove from start and end of the string
+      * @return a new string with removed leading and tailing matching f function
+      */
+    def dropWhileLeftRight(f: Char => Boolean): String = {
```
Comment (andreaTP) [inline src/main/scala/codacy/checkstyle/DocGenerator.scala]: ok, reading also the test this has to be implemented with a RegExp
